### PR TITLE
lang: funcs: Add more autogenerated functions

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -412,14 +412,10 @@ help: ## show this help screen
 		awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-20s\033[0m %s\n", $$1, $$2}'
 	@echo ''
 
-funcgen: lang/funcs/core/generated_funcs_test.go lang/funcs/core/generated_funcs.go
-
-lang/funcs/core/generated_funcs_test.go: lang/funcs/funcgen/*.go lang/funcs/core/funcgen.yaml lang/funcs/funcgen/templates/generated_funcs_test.go.tpl
-	@echo "Generating: funcs test..."
```
Comment (purpleidea) [inline Makefile]: Why did we nuke the tests?

File: lang/funcs/core/funcgen.yaml [hunk 2]
```diff
@@ -1,55 +1,25 @@
...
+packages:
+- name: html
+- name: math
+  exclude:
+    - Mod
+    - Pow
+    - Sqrt
+- name: math/rand
+  alias: rand
+  mgmtAlias: rand
```
Comment (purpleidea) [inline lang/funcs/core/funcgen.yaml]: Can we document all the different fields somewhere? Maybe with struct doc strings?

File: lang/funcs/funcgen/config.go [hunk 3]
```diff
@@ -24,7 +24,16 @@ import (
...
 
 type config struct {
-	Functions functions `yaml:"functions"`
+	Packages gopackages `yaml:"packages"`
+}
+
+type gopackages []gopackage
+
+type gopackage struct {
+	Name      string   `yaml:"name"`
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: We should docstring each of these fields. See here for an example: https://github.com/purpleidea/mgmt/blob/d5074871c78eacb4616127af074a7ede923d352c/engine/resources.go#L168

File: lang/funcs/funcgen/config.go [hunk 4]
```diff
@@ -36,23 +45,51 @@ type testarg struct {
 }
 
 type arg struct {
-	Name string `yaml:"name,omitempty"`
-	Type string `yaml:"type"`
+	Name  string `yaml:"name,omitempty"`
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: Docstrings please, see above.

File: lang/funcs/funcgen/config.go [hunk 5]
```diff
@@ -36,23 +45,51 @@ type testarg struct {
...
+	Value string `yaml:"value,omitempty"`
+	Type  string `yaml:"type"`
+}
+
+// GoType prints the go equivalent of a mcl type.
+func (obj *arg) GoType() string {
+	t := obj.Type
+	if t == "float" {
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: What about string vs str, etc?

File: lang/funcs/funcgen/config.go [hunk 6]
```diff
@@ -36,23 +45,51 @@ type testarg struct {
...
+		return "float64"
+	}
+	return t
 }
 
 // ToMcl prints the arg signature as expected by mcl.
 func (obj *arg) ToMcl() (string, error) {
+	if obj.Type == "bool" {
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: replace this with `switch obj.Type {` and `case`

File: lang/funcs/funcgen/config.go [hunk 7]
```diff
@@ -36,23 +45,51 @@ type testarg struct {
...
+			return fmt.Sprintf("%s float", obj.Name), nil
+		}
+		return types.TypeFloat.String(), nil
+	}
 	return "", fmt.Errorf("cannot convert %v to mcl", obj)
 }
 
 // ToGo prints the arg signature as expected by golang.
 func (obj *arg) ToGo() (string, error) {
+	if obj.Type == "float" {
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: `switch/case` please

File: lang/funcs/funcgen/func.go [hunk 8]
```diff
@@ -28,42 +28,34 @@ import (
...
 type function struct {
-	MgmtPackage string     `yaml:"mgmtPackage"`
-	MgmtName    string     `yaml:"mgmtName"`
-	Help        string     `yaml:"help"`
-	GoPackage   string     `yaml:"goPackage"`
-	GoFunc      string     `yaml:"goFunc"`
-	Args        []arg      `yaml:"args"`
-	Return      []arg      `yaml:"return"`
-	Tests       []functest `yaml:"tests"`
+	MgmtPackage  string    `yaml:"mgmtPackage"`
```
Comment (purpleidea) [inline lang/funcs/funcgen/func.go]: docstrings please so we know how to use this in the future.

File: lang/funcs/funcgen/pkg.go [hunk 9]
```diff
@@ -18,13 +18,25 @@
...
+	"regexp"
+	"strings"
 
+	"github.com/iancoleman/strcase"
 	yaml "gopkg.in/yaml.v2"
 )
 
+var (
+	validSignature = regexp.MustCompile(`^func (?P<name>[A-Z][a-zA-Z0-9]+)\((?P<args>([a-zA-Z]+( (string|bool|float64)|, )){0,})\) (?P<return>(string|float64|bool)|\((string|float64|bool), error\))$`)
+	excludedErr    = errors.New("Function is excluded")
```
Comment (purpleidea) [inline lang/funcs/funcgen/pkg.go]: `errExcluded` instead please, and error messages are not capitalized (test failure)

File: lang/funcs/funcgen/pkg.go [hunk 10]
```diff
@@ -37,9 +49,175 @@ func parsePkg(path, filename, templates string) error {
...
+		gopkgfuncs, err := gopackage.Parsefuncs()
+		if err != nil {
+			return funcs, err
+		}
+		funcs = append(funcs, gopkgfuncs...)
+	}
+	return funcs, nil
+}
+
+func (obj *gopackage) Parsefuncs() (functions, error) {
```
Comment (purpleidea) [inline lang/funcs/funcgen/pkg.go]: docstrings...

File: lang/funcs/funcgen/pkg.go [hunk 11]
```diff
@@ -37,9 +49,175 @@ func parsePkg(path, filename, templates string) error {
...
+				funcs = append(funcs, f)
+			}
+		}
+	}
+
+	return funcs, nil
+}
+
+func (obj *gopackage) parseFunctionLine(line string) (function, error) {
+
```
Comment (purpleidea) [inline lang/funcs/funcgen/pkg.go]: kill nl

File: lang/funcs/funcgen/pkg.go [hunk 12]
```diff
@@ -37,9 +49,175 @@ func parsePkg(path, filename, templates string) error {
...
+		mgmtpackage = obj.MgmtAlias
+	}
+
+	internalName := fmt.Sprintf("%s_%s", strcase.ToCamel(strings.ReplaceAll(obj.Name, "/", "_")), name)
+	help, err := obj.getHelp(name, internalName)
+	if err != nil {
+		return function{}, err
+	}
+
+	return function{
```
Comment (purpleidea) [inline lang/funcs/funcgen/pkg.go]: please switch all the structs to *function instead.

File: Makefile (hunk 1 above)
Comment (roidelapluie) [reply Makefile]: coz we can not autogenerate them anymore

File: lang/funcs/funcgen/config.go (hunk 5 above)
Comment (roidelapluie) [reply lang/funcs/funcgen/config.go]: string is correct

File: lang/funcs/funcgen/func.go (hunk 8 above)
Comment (roidelapluie) [reply lang/funcs/funcgen/func.go]: done

File: lang/funcs/funcgen/pkg.go (hunk 12 above)
Comment (roidelapluie) [reply lang/funcs/funcgen/pkg.go]: done

File: lang/funcs/funcgen/pkg.go (hunk 11 above)
Comment (roidelapluie) [reply lang/funcs/funcgen/pkg.go]: done

File: lang/funcs/funcgen/config.go (hunk 3 above)
Comment (roidelapluie) [reply lang/funcs/funcgen/config.go]: done

File: lang/funcs/core/funcgen.yaml (hunk 2 above)
Comment (roidelapluie) [reply lang/funcs/core/funcgen.yaml]: done

File: Makefile [hunk 13]
```diff
@@ -412,14 +412,10 @@ help: ## show this help screen
...
-	@go run lang/funcs/funcgen/*.go -templates lang/funcs/funcgen/templates/generated_funcs_test.go.tpl 2>/dev/null
+funcgen: lang/funcs/core/generated_funcs.go
 
 lang/funcs/core/generated_funcs.go: lang/funcs/funcgen/*.go lang/funcs/core/funcgen.yaml lang/funcs/funcgen/templates/generated_funcs.go.tpl
 	@echo "Generating: funcs..."
-	@go run lang/funcs/funcgen/*.go -templates lang/funcs/funcgen/templates/generated_funcs.go.tpl 2>/dev/null
```
Comment (purpleidea) [inline Makefile]: why get rid of ? 2>/dev/null

File: examples/lang/html.mcl [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import "html"
+import "fmt"
+
+$text1 = html.unescape_string("&lt;h1&gt;MGMT!&lt;/h1&gt;")
+$text2 = html.escape_string("Test & Re-Test\n")
+
```
Comment (purpleidea) [inline examples/lang/html.mcl]: kill 1nl please

File: examples/lang/html.mcl [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+
+file "/tmp/index.html" {
+    content => "${text1}${text2}",
```
Comment (purpleidea) [inline examples/lang/html.mcl]: tabs for indentation, here and everywhere in mcl

File: lang/funcs/funcgen/config.go [hunk 16]
```diff
@@ -24,45 +24,91 @@ import (
...
 
-type functions []function
+type gopackages []gopackage
 
-type testarg struct {
-	Name  string `yaml:"name,omitempty"`
-	Type  string `yaml:"type"`
-	Value string `yaml:"value"`
+type gopackage struct {
+	// Name is the name of the go package.
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: s/go/golang please

File: lang/funcs/funcgen/config.go [hunk 17]
```diff
@@ -24,45 +24,91 @@ import (
...
+	Name string `yaml:"name"`
+	// Alias is the alias of the package when imported in go.
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: s/go/golang please

File: lang/funcs/funcgen/config.go [hunk 18]
```diff
@@ -24,45 +24,91 @@ import (
...
+	// e.g. import rand "os.rand"
+	Alias string `yaml:"alias,omitempty"`
+	// MgmtAlias is the name of the package inside mcl.
+	MgmtAlias string `yaml:"mgmtAlias,omitempty"`
+	// Exclude is a list of go function name we do not want.
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: s/go/golang please

File: lang/funcs/funcgen/config.go [hunk 19]
```diff
@@ -24,45 +24,91 @@ import (
...
-	Type  string `yaml:"type"`
-	Value string `yaml:"value"`
+type gopackage struct {
+	// Name is the name of the go package.
+	Name string `yaml:"name"`
+	// Alias is the alias of the package when imported in go.
+	// e.g. import rand "os.rand"
+	Alias string `yaml:"alias,omitempty"`
+	// MgmtAlias is the name of the package inside mcl.
+	MgmtAlias string `yaml:"mgmtAlias,omitempty"`
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: s/MgmtAlias/MclAlias/ ?

File: lang/funcs/funcgen/config.go [hunk 20]
```diff
@@ -24,45 +24,91 @@ import (
...
 
+type functions []function
+
 type arg struct {
+	// Name is the name of the argument.
 	Name string `yaml:"name,omitempty"`
+	// Value is the value of the argument.
+	Value string `yaml:"value,omitempty"`
+	// Type is the type of the argument.
+	// Supported: string, float, bool.
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: not int?

File: lang/funcs/funcgen/config.go (hunk 5 above)
Comment (purpleidea) [reply lang/funcs/funcgen/config.go]: what does this mean?

File: lang/funcs/funcgen/config.go (hunk 6 above)
Comment (purpleidea) [reply lang/funcs/funcgen/config.go]: this was marked as resolved but not changed :(

File: lang/funcs/funcgen/config.go [hunk 21]
```diff
@@ -24,45 +24,91 @@ import (
...
 // ToTestInput prints the arg signature as expected by tests.
 func (obj *arg) ToTestInput() (string, error) {
-	if obj.Type == "string" {
+	switch obj.Type {
+	case "bool":
+		return fmt.Sprintf("&types.BoolValue{V: %s}", obj.Name), nil
+	case "float":
+		return fmt.Sprintf("&types.FloatValue{V: %s}", obj.Name), nil
+	case "string":
 		return fmt.Sprintf("&types.StrValue{V: %s}", obj.Name), nil
```
Comment (purpleidea) [inline lang/funcs/funcgen/config.go]: where is int?

File: lang/funcs/funcgen/func.go [hunk 22]
```diff
@@ -28,42 +28,48 @@ import (
...
-	Help        string     `yaml:"help"`
-	GoPackage   string     `yaml:"goPackage"`
-	GoFunc      string     `yaml:"goFunc"`
-	Args        []arg      `yaml:"args"`
-	Return      []arg      `yaml:"return"`
-	Tests       []functest `yaml:"tests"`
+	// MgmtName is the name of the package of the function in mcl.
+	MgmtPackage string `yaml:"mgmtPackage"`
+	// MgmtName is the name of the function in mcl.
+	MgmtName string `yaml:"mgmtName"`
```
Comment (purpleidea) [inline lang/funcs/funcgen/func.go]: MclName

File: lang/funcs/funcgen/func.go [hunk 23]
```diff
@@ -28,42 +28,48 @@ import (
...
+	// InternalName is the name used inside the templated file.
+	// Used to avoid clash between same functions from different packages.
+	InternalName string `yaml:"internalName"`
+	// Help is the docstring of the function, includung // and
```
Comment (purpleidea) [inline lang/funcs/funcgen/func.go]: typo

**Comment by roidelapluie** [comment]
@purpleidea added tests

**Comment by purpleidea** [comment]
Nested system imports are fixed, so you can now have `import "golang/whatever" and it will work. You do this by registering the function like so: `simple.ModuleRegister("golang"+"/"+"whatever", "some_function", &types.FuncValue{ ... })`

It also works for "golang/math/rand" type things as well.

This should now let you add the "golang" prefix to all these autogenerated things. LMK when patch is ready for new review.

**Comment by purpleidea** [comment]
@roidelapluie Reminder as requested.

**Comment by purpleidea** [comment]
Rebased to master and merged \o/
Only took six months ;)
Welcome back!

**Pull request closed without merging**